[3/3] DMA Move API: Introduce DMA buffer objects

Dominaezzz opened the pull request:
## Thank you for your contribution!

We appreciate the time and effort you've put into this pull request.
To help us review it efficiently, please ensure you've gone through the following checklist:

### Submission Checklist 📝
- [x] I have updated existing examples or added new ones (if applicable).
- [x] I have used `cargo xtask fmt-packages` command to ensure that all changed code is formatted correctly.
- [x] My changes were added to the [`CHANGELOG.md`](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/CHANGELOG.md) in the **_proper_** section.
- [x] My changes are in accordance to the [esp-rs API guidelines](https://github.com/esp-rs/esp-hal/blob/main/API-GUIDELINES.md)

#### Extra:
- [x] I have read the [CONTRIBUTING.md guide](https://github.com/esp-rs/esp-hal/blob/main/CONTRIBUTING.md) and followed its instructions.

### Pull Request Details 📖

#### Description
The (not so) final PR in the DMA move API saga in #1785 (previously #1512).
(Also closes #1798 by removing the struct in question).
The code in this PR could do with several enhancements but I've left them for a future PR (which is in the works) as this PR is big and catastrophic enough. I didn't want to go changing too much in one go, making it tough to review.

This PR does a couple things:
- Introduces `DmaTxBuf` struct. Which is a buffer and descriptor combo used for sending data with the DMA.
- Introduces `DmaRxBuf` struct. Which is a buffer and descriptor combo used for receiving data with the DMA.
- Introduces `DmaTxRxBuf` struct. Which is a buffer and descriptor combo used for full duplex DMA.
- Migrates the `SpiDma` struct to a move only API. (This is to ensure soundness in the presence of `mem::forget`).
- Introduces a `SpiDmaTransfer` struct to represent an in-progress spi dma transfer.
- Provides a `SpiDmaBus` and `SpiDmaAsyncBus` structs which wrap the `SpiDma` driver (and some dma bufs) to provide a convenient API that works with slices and doesn't require moving things around.
- Added HIL tests

I'm happy to rename things. :slightly_smiling_face: 

Future planned enhancements:
- Improving error handling/propagation.
  - There's a handful of errors that can occur that the existing driver doesn't expose and exposing them in the code as-is is awkward due to fact that the current driver supports asymmetric transfers, which the hardware reports errors for, because as far as the SPI peripheral is concerned, all transfers are symmetric. In a future PR, I'll remove support for asymmetric transfers in the core move based API, but leave it in the `SpiDmaBus` layer. Then error propagation can be implemented.
  - The semantics need to be decided for when the DMA transfer fails before the SPI transfer is done. SPI transfers can't be cancelled according to the TRM, so do we just wait until the SPI peripheral to finish sending garbage? Maybe setting SPI_USR to zero would actually stop the transfer but it's not documented.... 
- Move/split/refactor the SPI specific `SpiDmaTransfer` struct to the dma module as a generic `DmaTransfer` (name pending) struct. I didn't do this in this PR due to error propagation issues I described above. Asymmetric transfers as they are need special care.
- Introduce some traits for `DmaTxBuf`, `DmaRxBuf` and `DmaTxRxBuf` to implement as I highly anticipate other buffer types that enforce things like alignment and burst mode, or specific buffer type for external memory, or a buf type that temporarily combines multiple bufs into one buf, etc.
- Where the `SpiDmaBus` and `SpiDmaAsyncBus` does chunking, the implementation can be enhanced to fill up a dma buf with the next chunk whilst the previous chunk is in flight, eliminating the latency gaps between each chunk. Haven't done this here as this PR is big enough.
- Allow `SpiDmaBus` and `SpiDmaAsyncBus` to work in half duplex mode. Doing this right now leads to too much duplication, I'm exploring alternatives in https://github.com/esp-rs/esp-hal/discussions/1853 .
- Allow quick conversions between `DmaTxBuf` and `DmaRxBuf` (by just moving the descriptors and buffer), as this will make it trivial to implement zero-copy peripheral to peripheral transfer. Imagine receiving a JPEG from LCD_CAM into a `DmaRxBuf`, then convert it into a `DmaTxBuf` and give it to the AES to encrypt it before storing it to an SD card.
- Add some convenience macros to create dma buffers.

#### Testing
I've ran the related HIL tests and examples on an S3 and C6.

bjoernQ commented on esp-hal/src/spi/master.rs:
```diff
@@ -1449,270 +1387,486 @@ pub mod dma {
-                    if idx >= write.len() as isize && idx >= read.len() as isize {
-                        break;
-                    }
+        impl<'d, T, C> embedded_hal_async::spi::SpiBus for SpiDmaAsyncBus<'d, T, C>
```
I know it wasn't like that before but in general we want to provide an inherent API and have ecosystem traits optionally implemented - since you are touching all of this anyways it's probably a chance to do it now?

bjoernQ replied:
But not really required in this PR - I understood you planned follow up PRs already

Dominaezzz replied:
Sure, I can do this in the PR that removes asymmetric transfers from the core API. (This makes it a bit easier for me)

We also need to decide what the inherent API should look like, copying the e-hal trait seems lazy to me but I can't think of anything else

bjoernQ replied:
> ... copying the e-hal trait seems lazy to me but I can't think of anything else

I guess that's good enough 👍

bjoernQ commented:
I guess `FlashSafeDma` could be removed with these changes

bjoernQ commented:
Apparently, this breaks the `qspi_flash` example on ESP32-C6 for me (accessing a GD25Q64C)

On current main (`81f3776aa79d5b6d0098ecf499abaf03e0a6ef56`) it works fine and prints the expected output. On this branch it just freezes

bjoernQ commented:
Seems like `dma_write` is somehow broken - changing `spi_loopback_dma` to use `dma_write` instead of `dma_transfer` works on current main but freezes with this branch

Dominaezzz commented:
> Apparently, this breaks the `qspi_flash` example on ESP32-C6 for me (accessing a GD25Q64C)
> 
> On current main (`81f3776aa79d5b6d0098ecf499abaf03e0a6ef56`) it works fine and prints the expected output. On this branch it just freezes

I sadly don't have that chip to test. Does it freeze in the loop or in the first few commands?

bjoernQ commented:
> > Apparently, this breaks the `qspi_flash` example on ESP32-C6 for me (accessing a GD25Q64C)
> > On current main (`81f3776aa79d5b6d0098ecf499abaf03e0a6ef56`) it works fine and prints the expected output. On this branch it just freezes
> 
> I sadly don't have that chip to test. Does it freeze in the loop or in the first few commands?

Seems like it freezes in the first call to `wait` - you can also run the example without anything attached (maybe a logic analyzer)

Dominaezzz commented:
Whilst I made a silly mistake when creating the `SpiDmaTransfer` struct, I also found an existing bug in SpiDma driver.
`write`s with a buffer of length zero will return a `DmaTransferRx` who's `is_done()` will never return true, even though `wait()` will quickly return.
Long story short, the channel's EOF mode (emit EOF interrupt when descriptor data is removed from the DMA FIFO, rather than when the descriptor data enters the DMA FIFO) means the TX channel never completes.

I've worked around this problem for now but I'm not sure what the long term solution for this should be. Perhaps transfer with data length zero shouldn't be allowed? Not sure.

bjoernQ commented:
Thanks!

Unfortunately, on this branch
- `cargo xtask run-example examples esp32c6 qspi_flash`  still freezes
- `cargo xtask run-example examples esp32c6 spi_loopback_dma`  freezes when changing the example to just do a write instead of transfer (same for just `dma_read`)

Both is working on `main` currently

Dominaezzz commented:
The workaround didn't work out, the GDMA channel isn't behaving like I expect so I've reverted to the behaviour on `main` for now.

bjoernQ commented:
Thanks! The `spi_loopback_dma` works fine again also when using `dma_write`/`dma_read`

The `qspi_flash` example (when using an SPI flash chip) also doesn't freeze anymore but there seem to be still issues compared to `main`

Output on main (as expected):
```
[48, 65, 6c, 6c, 6f, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff]
Hello!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!................................................................
[48, 65, 6c, 6c, 6f, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff]
Hello!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!................................................................
```

Output on this branch:
```
[48, 0, 65, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff]
H.elllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllll................................................................
[48, 0, 65, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, 6c, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff]
H.elllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllll................................................................
```

I know you don't have that SPI flash chip but maybe a logic-analyzer will reveal the problem, too?

It also seems to be a problem with `write` since running the example from main and commenting out writing on this branch reads the correct data apparently

Dominaezzz commented:
Thanks! Yeah I can see it in my logical analyzer. It appears the first write works as expected but subsequent writes don't.

Dominaezzz commented:
Well this is annoying, the issue is due to timing.... (Originally found in https://github.com/esp-rs/esp-hal/issues/489 it seems!)

The ESP32 chips insist on having a (undocumented) delay between starting the DMA and starting the peripheral.

Now that I've moved the descriptor setup/initialization out of the SPI driver, the delay from doing it is gone, which means the DMA doesn't have enough time to fill the SPI async fifo before the SPI peripheral starts sending data.

I was aware that LCD_CAM had this issue but it seems it's a general thing.

Note: The reason `dma_transfer` works and `dma_write` doesn't, is because the RX channel is setup after the TX one, which functions as a delay... :slightly_smiling_face:.

I've added a delay to fix this but I'm open to alternatives

Dominaezzz commented on esp-hal/src/soc/mod.rs:
```diff
@@ -72,6 +72,13 @@ pub(crate) fn is_valid_ram_address(address: u32) -> bool {
 }
 
+#[allow(unused)]
+pub(crate) fn is_slice_in_dram<T>(slice: &[T]) -> bool {
```
I can't imagine a user creating such a slice without using `unsafe`. (I'm not sure whether or not it's unsound to do so either).

bjoernQ reviewed (approved):
Thanks! - Code looks good to me and everything works fine (again)

liebman commented on esp-hal/src/dma/mod.rs:
```diff
@@ -1840,6 +1878,519 @@ where
+    /// Only DRAM is supported.
+    pub fn new(
+        descriptors: &'static mut [DmaDescriptor],
+        buffer: &'static mut [u8],
```
Could this be something that implements WriteBuffer instead?  For instance  A frame buffer that us made from a `&'static mut [u16]` or event `&'static mut [Entry]`.

ProfFan replied:
`embedded_dma` seems to have been abandoned, just my 5 cents

Dominaezzz replied:
As one of the further enhancements I'm planning to make, I'm going to make a trait for these buffer objects and this will make it easier to throw in customisations like this.

(This will likely be the next PR I make)

I want to keep this PR simple for now. The merge conflicts are already killing my motivation.

liebman replied:
Yes but these were re-implemented as `esp_hal::dma::ReadBuffer` and `esp_hal::dma::WriteBuffer`

liebman replied:
Let me know if I can help - I have a driver for HUB75 displays where I Implement `ReadBuffer`.

liebman commented on esp-hal/src/dma/mod.rs:
```diff
@@ -1840,6 +1878,519 @@ where
+    /// Only DRAM is supported.
+    pub fn new(
+        descriptors: &'static mut [DmaDescriptor],
+        buffer: &'static mut [u8],
```
Same as above but implementing `ReadBuffer`?

bugadani commented on esp-hal/src/lib.rs:
````diff
@@ -570,61 +570,6 @@ mod critical_section_impl {
-///
-/// spi.write(&ARRAY_IN_FLASH[..]).unwrap(); // success
-/// ```
-pub struct FlashSafeDma<T, const SIZE: usize> {
````
Removing this needs to be mentioned in the changelog

Dominaezzz replied:
Ah yeah fair enough. So two entries in the changelog file with the same PR number? One in Add and one in Remove?

bugadani replied:
Sure, why not, there's already one entry in Add and one in Change :) Just a mention so that the API doesn't disappear without a trace.

MabezDev reviewed (commented):
Firstly, I'm very sorry for taking so long to get to this! I really appreciate the effort your putting in and I will make it a priority to review your PRs in a more timely fashion.

Overall, the PR, and particularly he SPI driver improvements are really nice. I am also glad that FlashSafeDma is now dead :D. I do have a couple of concerns which I think are worth discussing.

I think it's unlikely that a new user is going to want to use the move-based API initially. I think that the most common path would be to use `SpiDmaBus` or `SpiDmaAsyncBus`which begs the question, should this be the default behaviour of `SpiDma`?

Perhaps we could have a `with_dma()` call that includes a set of descriptors (how it works before this PR), and a seperate `with_dma_split_descriptors` (I'm sure there is a better way to name this) that gives you a move based SPI where you passed the descriptors to allow for queuing etc.

What do you think?

Dominaezzz commented:
Thanks for having a look. (I appreciate you've been on leave and also ill for a bit)

Yes I see where you're coming from, make the common case the default path.
There's the complication that `SpiDma(Async)Bus` doesn't support half duplex but I should be able to work around that with some where clauses.

Dominaezzz commented:
I couldn't figure out the where clauses :sweat_smile:. I've added a `with_buffers(...)` method to make it easy to go the default path.

```Rust
.with_dma(channel)
.with_buffers(dma_tx_buf, dma_rx_buf);
```

MabezDev reviewed (approved):
> I couldn't figure out the where clauses 😅. I've added a `with_buffers(...)` method to make it easy to go the default path.
> 
> ```rust
> .with_dma(channel)
> .with_buffers(dma_tx_buf, dma_rx_buf);
> ```

I think this works well!

This LGTM, thanks for sticking with me, I think the end result is really nice! Looking forward to the future PRs.

bugadani commented on esp-hal/src/dma/mod.rs:
```diff
@@ -1840,6 +1878,519 @@ where
+    }
+
+    /// Returns the buf as a mutable slice than can be written.
+    pub fn as_mut_slice(&mut self) -> &mut [u8] {
```
In its current form this function is a footgun. It gives access to the whole buffer, but it does not update the length.

Should we
- ... set the maximum length when this is called?
- ... only return a subset of the buffer, controlled by `set_length`?
- ... document that the user should probably call `set_length`?

Dominaezzz replied:
I agree that this is confusing, which is why I didn't implement `Deref`/`AsRef`/`Borrow`.

I know I definitely want users to be able to access the whole buffer regardless of length somehow, how this is done I'm not too fussed about.

Definitely not the first option at least. Since there will be no way to edit the buffer without changing its length.

This is issue worthy



> In its current form this function is a footgun.

I hope you didn't shoot your foot recently :sweat_smile:

MabezDev replied:
Some more options:

* A closure based API where the user has to return the number of bytes "used".
* We remove `as_mut_slice` and rely on `fill` only.
* We consider DmaTxBuf etc short lived, slice manipulation should happen outside before passing it into DmaTxBuf
